lgc/patch/PatchResourceCollect.cpp

```diff
@@ -283,6 +260,130 @@ void PatchResourceCollect::setNggControl() {
+
+  if (!usePosition)
+    return false; // No position export
+
+  // Find position export call
+  std::string posCallName = lgcName::OutputExportBuiltIn;
+  posCallName += PipelineState::getBuiltInName(BuiltInPosition);
+  auto callStage = hasGs ? ShaderStageGeometry : (hasTs ? ShaderStageTessEval : ShaderStageVertex);
+
+  CallInst *posCall = nullptr;
+  for (Function &func : *module) {
+    if (func.getName().startswith(posCallName)) {
+      for (User *user : func.users()) {
+        auto call = cast<CallInst>(user);
+        if (m_pipelineShaders->getShaderStage(call->getFunction()) == callStage) {
+          posCall = call;
+          break;
+        }
+      }
+
+      if (posCall) // Already find the call
+        break;
+    }
+  }
+  assert(posCall); // Position export must exist
+
+  // Check position value, disable NGG culling if it is produced by instructions
+  auto posValue = posCall->getArgOperand(posCall->getNumArgOperands() - 1); // Last argument is position value
+  if (!isa<Instruction>(posValue))
+    return false;
+
+  // For GS, we can safely enable NGG culling here; for non-GS, still need further analysis
+  if (hasGs)
+    return true;
+
+  // Check if the sources of position value are eventually produced by vertex-based inputs loaded from memory
+  SmallVector<Instruction *, 32> posSources;
+  posSources.push_back(cast<Instruction>(posValue));
+
+  while (!posSources.empty()) {
+    auto posSource = posSources.back();
+    posSources.pop_back();
+
+    // Pop up a source and check its operands
+    for (auto &operand : posSource->operands()) {
+      if (isa<Instruction>(operand)) {
+        // Only check instructions
+        auto inst = cast<Instruction>(operand);
+        if (isa<CallInst>(inst)) {
+          // Further check calls
+          auto calleeName = cast<CallInst>(inst)->getCalledFunction()->getName();
+          // Generic inputs are always vertex-based
+          if (calleeName.startswith(lgcName::InputImportGeneric))
+            return true;
+          // If tessellation, we have to further check certain built-in inputs that are vertex-based
+          if (hasTs && calleeName.startswith(lgcName::InputImportBuiltIn)) {
+            auto builtInName = calleeName.slice(strlen(lgcName::InputImportBuiltIn), StringRef::npos);
+            if (builtInName.startswith(PipelineState::getBuiltInName(BuiltInPosition)) ||
+                builtInName.startswith(PipelineState::getBuiltInName(BuiltInPointSize)) ||
+                builtInName.startswith(PipelineState::getBuiltInName(BuiltInClipDistance)) ||
+                builtInName.startswith(PipelineState::getBuiltInName(BuiltInCullDistance)) ||
+                builtInName.startswith(PipelineState::getBuiltInName(BuiltInTessLevelOuter)) ||
+                builtInName.startswith(PipelineState::getBuiltInName(BuiltInTessLevelInner)))
+              return true;
+          }
+        }
+        posSources.push_back(inst);
+      }
+    }
+  }
```

Comment:
Some remarks here:

1. The complexity of the search can easily escalate if the data flow leading to the position export is not a tree. Actually, if part of the calculation leading to the position export is a loop, this code will enter an infinite loop (because it will keep going over the phi nodes of the loop header). We need to remember the already-visited instructions in a set so we don't visit them multiple times.

2. What about vertex shaders that don't use vertex arrays but instead load from an SSBO or texel buffer based on the BulitInVertexIndex? I don't think we should disable culling for those, considering that I've heard some ISVs are experimenting with moving away from vertex arrays in order to do custom compaction of data.

3. Finally, I have to say that despite our internal discussions, I'm still not convinced that the compile time spent on this is worth the hassle, and to be perfectly honest with you, I'd prefer it if we didn't do this until more concrete data is shown. That said, I'm not going to stand in the way of adding it, but I would ask you to please move this analysis into a separate method, so that switching it on and off becomes easier in the future.

Reply:
> The complexity of the search can easily escalate if the data flow leading to the position export is not a tree. Actually, if part of the calculation leading to the position export is a loop, this code will enter an infinite loop (because it will keep going over the phi nodes of the loop header). We need to remember the already-visited instructions in a set so we don't visit them multiple times.

Thanks for pointing this out. I do consider such cases.

> What about vertex shaders that don't use vertex arrays but instead load from an SSBO or texel buffer based on the BulitInVertexIndex? I don't think we should disable culling for those, considering that I've heard some ISVs are experimenting with moving away from vertex arrays in order to do custom compaction of data.

Yes, this case is valid for culling. We do have to allow it that is checking a buffer memory loading with gl_VertexIndex or something similar as the indices in access chain.

> Finally, I have to say that despite our internal discussions, I'm still not convinced that the compile time spent on this is worth the hassle, and to be perfectly honest with you, I'd prefer it if we didn't do this until more concrete data is shown. That said, I'm not going to stand in the way of adding it, but I would ask you to please move this analysis into a separate method, so that switching it on and off becomes easier in the future.

That's fine. I will simplify the implementation until we have concrete experimental data that strongly disprove the culling is effective and should be disable in such cases.